website/docs/r/iam_role.html.markdown

```diff
@@ -91,6 +65,166 @@ resource "aws_iam_role" "instance" {
 }
 ```
 
+### Example of Exclusive Inline Policies
+
+This example creates an IAM role with two inline IAM policies. If someone adds another inline policy out-of-band, on the next apply, Terraform will remove that policy. If someone deletes these policies out-of-band, Terraform will recreate them.
+
+```hcl
+resource "aws_iam_role" "example" {
+  name               = "yak_role"
+  assume_role_policy = data.aws_iam_policy_document.instance_assume_role_policy.json # (not shown)
+
+  inline_policy {
+    name = "my_inline_policy"
+
+    policy = <<EOF
+{
+  "Statement": [
+    {
+      "Action": [
+        "ec2:Describe*"
+      ],
+      "Effect": "Allow",
+      "Resource": "*"
+    }
+  ],
+  "Version": "2012-10-17"
+}
+EOF
+  }
+
+  inline_policy {
+    name   = "policy-8675309"
+    policy = data.aws_iam_policy_document.inline_policy.json
+  }
+}
+
+data "aws_iam_policy_document" "inline_policy" {
+  statement {
+    actions   = ["ec2:DescribeAccountAttributes"]
+    resources = ["*"]
+  }
+}
+```
+
+### Example of Removing Inline Policies
+
+This example creates an IAM role with what appears to be empty IAM `inline_policy` argument instead of using `inline_policy` as a configuration block. The result is that if someone were to add an inline policy out-of-band, on the next apply, Terraform will remove that policy.
+
+
+```hcl
+resource "aws_iam_role" "example" {
+  name               = "yak_role"
+  assume_role_policy = data.aws_iam_policy_document.instance_assume_role_policy.json # (not shown)
+
+  inline_policy {}
```

Comment:
Curious... Does this work with dynamic blocks? Say `var.inline_policies` is a list of objects. If the list is empty, does this remove all inline policies?

```
  dynamic inline_policy {
    for_each = policy in var.inline_policies
    content {
      name   = policy.value.name
      policy = policy.value.policy
    }
  }
```

Or, how might one do that?

Reply:
It _should_ be equivalent. `inline_policy {}` internally becomes a length 1 set with the one element being a map, which has name and policy fields, but they are empty. That's the key indicator that policies should be removed since it produces a diff if there are any policies found in the read.

Reply:
> `inline_policy {}` internally becomes a length 1 set with the one element being a map, which has name and policy fields, but they are empty

you mean these are equivalent?

```
  inline_policy {}
```

```
  inline_policy {
    name   = null
    policy = null
  }

```

If the dynamic block doesn't "just work" then I can probably figure out how to force that in the `for_each` expression using a condition...

Reply:
I verified that these two configs behave the same (out of band policies are removed):

```hcl
locals {
  inline_policies = []
}

resource "aws_iam_role" "test" {
  name = %[1]q

  assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Action": "sts:AssumeRole",
      "Principal": {
        "Service": "ec2.amazonaws.com"
      },
      "Effect": "Allow",
      "Sid": ""
    }
  ]
}
EOF

  dynamic "inline_policy" {
    for_each = local.inline_policies
    content {
      name = inline_policy.value["name"]
      policy = inline_policy.value["policy"]
    }
  }
}
```

```hcl
resource "aws_iam_role" "test" {
  name = %[1]q

  assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Action": "sts:AssumeRole",
      "Principal": {
        "Service": "ec2.amazonaws.com"
      },
      "Effect": "Allow",
      "Sid": ""
    }
  ]
}
EOF

  inline_policy {}
}
```

Reply:
Oh, fantastic, that's awesome! Thanks!

Reply:
> you mean these are equivalent?

In this case, yes. I imagine that in some cases validation would prevent you from literally writing `name = null` even if internally it represented it that way. But, I verified that it appears as empty and causes out-of-band policies to be removed using 1. the plain empty, 2. the dynamic empty, or 3. the config block with null arguments. Equivalent syntax should produce equivalent internal representation and I'm glad to see that's the case here.

Reply:
FYI, this is why I had to change `name` and `policy` to `Optional`. It will still error if they are not provided but I'm relying on the API for the moment for that.